pkg/crds/crds.go

```diff
@@ -0,0 +1,160 @@
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crds
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/types"
+	"strings"
+)
+
+var cephResources = []string{
+	"cephblockpoolradosnamespaces",
+	"cephblockpools",
+	"cephbucketnotifications",
+	"cephbuckettopics",
+	"cephclients",
+	"cephclusters",
+	"cephcosidrivers",
+	"cephfilesystemmirrors",
+	"cephfilesystems",
+	"cephfilesystemsubvolumegroup",
+	"cephnfses",
+	"cephobjectrealms",
+	"cephobjectstores",
+	"cephobjectstoreusers",
+	"cephobjectzonegroups",
+	"cephobjectzones",
+	"cephrbdmirrors",
+}
+
+const (
+	cephRookIoGroup          = "ceph.rook.io"
+	cephRookResourcesVersion = "v1"
+)
+
+const (
+	CephResourceKindCephCluster = "CephCluster"
+)
+
+var (
+	errorResourceNotFound         = errors.New("the server could not find the requested resource")
+	clusterResourcePatchFinalizer = map[string]interface{}{
+		"spec": map[string]interface{}{
+			"cleanupPolicy": map[string]string{
+				"confirmation": "yes-really-destroy-data",
+			},
+		},
+	}
+)
+
+func DeleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) {
+	err := deleteCustomResources(ctx, clientsets, clusterNamespace)
+	if err != nil {
+		logging.Fatal(err)
+	}
+	logging.Info("done")
+}
+
+func deleteCustomResources(ctx context.Context, clientsets k8sutil.ClientsetsInterface, clusterNamespace string) error {
+	for _, resource := range cephResources {
+		logging.Info(fmt.Sprintf("Getting resources kind %s", resource))
+		items, err := clientsets.ListResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace)
+		if err != nil {
+			if strings.Contains(err.Error(), errorResourceNotFound.Error()) {
+				logging.Warning("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			// log instead of return
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Warning("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if err.Error() == "the server could not find the requested resource" {
+					logging.Info(err.Error())
```

Comment:
Is this an info level because there were no resources of this type found? How can this happen if we are iterating through the resources in the `for` loop? Seems like this should be logged as an error, with more details to indicate which resource failed.

Reply:
the loops contains all these resources

```go
var cephResources = []string{
	"cephblockpoolradosnamespaces",
	"cephblockpools",
	"cephbucketnotifications",
	"cephbuckettopics",
	"cephclients",
	"cephclusters",
	"cephcosidrivers",
	"cephfilesystemmirrors",
	"cephfilesystems",
	"cephfilesystemsubvolumegroup",
	"cephnfses",
	"cephobjectrealms",
	"cephobjectstores",
	"cephobjectstoreusers",
	"cephobjectzonegroups",
	"cephobjectzones",
	"cephrbdmirrors",
}
```

I added this validation to stop the process and log the error if something goes wrong when we are processing the resources

```go
if err != nil {
  if k8sErrors.IsNotFound(err) {
	  logging.Warning(err.Error())
	  continue
  }
  return err
}
```

By other side if the error is the type not found, this just will be logged and display a warning. I added this validation to avoid stop prematurely the process if some resource does not exists.

This happens continuously on minikube, even after a fresh installation.

Locally I installed minikube and setup rook

```bash
$minikube start --disk-size=40g --extra-disks=1 --driver kvm2
😄  minikube v1.30.1 on Fedora 38
✨  Using the kvm2 driver based on user configuration
👍  Starting control plane node minikube in cluster minikube
🔥  Creating kvm2 VM (CPUs=8, Memory=16192MB, Disk=40960MB) ...
🐳  Preparing Kubernetes v1.26.3 on Docker 20.10.23 ...
    ▪ Generating certificates and keys ...
    ▪ Booting up control plane ...
    ▪ Configuring RBAC rules ...
🔗  Configuring bridge CNI (Container Networking Interface) ...
    ▪ Using image gcr.io/k8s-minikube/storage-provisioner:v5
🌟  Enabled addons: storage-provisioner, default-storageclass
🔎  Verifying Kubernetes components...
🏄  Done! kubectl is now configured to use "minikube" cluster and "default" namespace by default

$ kubectl get all -n rook-ceph
NAME                                                READY   STATUS      RESTARTS   AGE
pod/csi-cephfsplugin-bmmlj                          2/2     Running     0          16m
pod/csi-cephfsplugin-provisioner-668dfcf95b-xhnlj   5/5     Running     0          16m
pod/csi-rbdplugin-ls27h                             2/2     Running     0          16m
pod/csi-rbdplugin-provisioner-5b78f67bbb-s9z62      5/5     Running     0          16m
pod/rook-ceph-exporter-minikube-5f7cf76bdf-ttvmh    1/1     Running     0          15m
pod/rook-ceph-mgr-a-596b878858-dlq9c                1/1     Running     0          16m
pod/rook-ceph-mon-a-dd57c4f84-797fq                 1/1     Running     0          16m
pod/rook-ceph-operator-59459dd8fc-2rbr8             1/1     Running     0          17m
pod/rook-ceph-osd-0-595d44574c-qsrg5                1/1     Running     0          15m
pod/rook-ceph-osd-prepare-minikube-b97sg            0/1     Completed   0          15m

NAME                              TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)             AGE
service/rook-ceph-exporter        ClusterIP   10.101.230.202   <none>        9926/TCP            16m
service/rook-ceph-mgr             ClusterIP   10.97.177.143    <none>        9283/TCP            15m
service/rook-ceph-mgr-dashboard   ClusterIP   10.102.22.166    <none>        7000/TCP            15m
service/rook-ceph-mon-a           ClusterIP   10.100.0.85      <none>        6789/TCP,3300/TCP   16m

NAME                              DESIRED   CURRENT   READY   UP-TO-DATE   AVAILABLE   NODE SELECTOR   AGE
daemonset.apps/csi-cephfsplugin   1         1         1       1            1           <none>          16m
daemonset.apps/csi-rbdplugin      1         1         1       1            1           <none>          16m

NAME                                           READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/csi-cephfsplugin-provisioner   1/1     1            1           16m
deployment.apps/csi-rbdplugin-provisioner      1/1     1            1           16m
deployment.apps/rook-ceph-exporter-minikube    1/1     1            1           16m
deployment.apps/rook-ceph-mgr-a                1/1     1            1           16m
deployment.apps/rook-ceph-mon-a                1/1     1            1           16m
deployment.apps/rook-ceph-operator             1/1     1            1           17m
deployment.apps/rook-ceph-osd-0                1/1     1            1           15m

NAME                                                      DESIRED   CURRENT   READY   AGE
replicaset.apps/csi-cephfsplugin-provisioner-668dfcf95b   1         1         1       16m
replicaset.apps/csi-rbdplugin-provisioner-5b78f67bbb      1         1         1       16m
replicaset.apps/rook-ceph-exporter-minikube-5f7cf76bdf    1         1         1       15m
replicaset.apps/rook-ceph-exporter-minikube-7556d677c5    0         0         0       16m
replicaset.apps/rook-ceph-mgr-a-596b878858                1         1         1       16m
replicaset.apps/rook-ceph-mon-a-dd57c4f84                 1         1         1       16m
replicaset.apps/rook-ceph-operator-59459dd8fc             1         1         1       17m
replicaset.apps/rook-ceph-osd-0-595d44574c                1         1         1       15m

NAME                                       COMPLETIONS   DURATION   AGE
job.batch/rook-ceph-osd-prepare-minikube   1/1           8s         15m

```

And this is the output after run the command 
```bash
$go run . destroy-cluster
Warning: Are you sure you want to destroy the cluster in namespace "rook-ceph"? If absolutely certain, enter: yes-really-destroy-cluster
yes-really-destroy-cluster
Info: proceeding
Info: Getting resources kind cephblockpoolradosnamespaces
Warning: resource cephblockpoolradosnamespaces was not found on the cluster
Info: Getting resources kind cephblockpools
Info: removing resource cephblockpools: builtin-mgr
Info: resource "builtin-mgr" is not yet deleted, applying patch to remove finalizer...
Warning: cephblockpools.ceph.rook.io "builtin-mgr" not found
Info: Getting resources kind cephbucketnotifications
Warning: resource cephbucketnotifications was not found on the cluster
Info: Getting resources kind cephbuckettopics
Warning: resource cephbuckettopics was not found on the cluster
Info: Getting resources kind cephclients
Warning: resource cephclients was not found on the cluster
Info: Getting resources kind cephclusters
Info: removing resource cephclusters: my-cluster
Info: resource "my-cluster" is not yet deleted, applying patch to remove finalizer...
Warning: cephclusters.ceph.rook.io "my-cluster" not found
Info: Getting resources kind cephcosidrivers
Warning: resource cephcosidrivers was not found on the cluster
Info: Getting resources kind cephfilesystemmirrors
Warning: resource cephfilesystemmirrors was not found on the cluster
Info: Getting resources kind cephfilesystems
Warning: resource cephfilesystems was not found on the cluster
Info: Getting resources kind cephfilesystemsubvolumegroup
Warning: the server could not find the requested resource: cephfilesystemsubvolumegroup
Info: Getting resources kind cephnfses
Warning: resource cephnfses was not found on the cluster
Info: Getting resources kind cephobjectrealms
Warning: resource cephobjectrealms was not found on the cluster
Info: Getting resources kind cephobjectstores
Warning: resource cephobjectstores was not found on the cluster
Info: Getting resources kind cephobjectstoreusers
Warning: resource cephobjectstoreusers was not found on the cluster
Info: Getting resources kind cephobjectzonegroups
Warning: resource cephobjectzonegroups was not found on the cluster
Info: Getting resources kind cephobjectzones
Warning: resource cephobjectzones was not found on the cluster
Info: Getting resources kind cephrbdmirrors
Warning: resource cephrbdmirrors was not found on the cluster
Info: done
$
```

I think we should not to stop the process if some resource was not found otherwise this process never will finish.

What do you think @travisn ?